### DISCUSSION TITLE
Hello!

### DISCUSSION THREAD

**Discussion by baggiponte** [python-rope/rope#670] [General]
Hi! I am a Junior ML Engineer based at my terminal. Of course, I use neovim 💘

While my setup with pyright works fine, I do not like how microsoft kept the more advanced functions in the proprietary pylance. I believe that rope is going in the right direction (especially since it's written in python) and I would like to offer my help in advancing the project.

Unfortunately, I am a total newbie in this sense. @lieryan suggested to work on the docs, and that's where I will start. My end goal would be to aid with the refactoring capabilities ("move this function/variable/class to another module and automatically update the imports") and neovim integration via `nvim-cmp`.

Please feel free to come up with other recommendations/priorities/guidance!

**Comment by edreamleo** [comment]
@baggiponte Hello! I am creating a Theory of Operation document for Rope's type inference. Perhaps our interests are complementary.

The [ekr-fork-rope2](https://github.com/edreamleo/ekr-fork-rope2/tree/devel) fork contains my investigations including:
- A simplified class hierarchy.
- New mypy annotations.
- A draft [theory.rst](https://github.com/edreamleo/ekr-fork-rope2/blob/devel/docs/theory.rst) document.

Please feel free to ask questions here about Rope. We may be able to help each other learn Rope's internals.

**Reply by baggiponte** [reply]
Thank you for your welcome! What is exactly a theory of operation? I am interested in the LSP theory, where should I start studying it from?

**Reply by lieryan** [reply]
Edward is studying the type inference in Rope and doing great work in understanding and documenting how that works. Hopefully, this can end up with improve Rope to actually understand type annotations as our support for rope making use of type hints is currently is a bit lacking.

If you want to understand how LSP works, the most important to understand would be the [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/). [pylsp-rope](https://github.com/python-rope/pylsp-rope) is the project that implements the LSP part of Rope for python-lsp-server, though the actual meat of the LSP implementation is actually on the [python-lsp-server](https://github.com/python-lsp/python-lsp-server) itself. 

An LSP is basically just a JSON-RPC server. The editor talks to the server usually via either stdin/stdout or tcp socket.

**Reply by baggiponte** [reply]
Okai, so to be clear `rope` aims to be a refactoring library and to make its features available inside an LSP we can use the pylsp rope plugin with python-lsp-server. Where does the type inference fit in this picture? First thing that comes to mind is to make smart completions available inside your IDE - say from an LSP, am I right?

**Reply by lieryan** [reply]
Type inference is a rather loose description of what rope does, it have some parallels to type inferrence in systems like mypy, but it also isn't really the same thing. Basically, rope needs to know whether a symbol that appears in a code are the same as another symbol, for example for a rename operation. It does some analysis to determine the type of a variable, what attributes/methods they have, etc.

**Reply by edreamleo** [reply]
@lieryan Many thanks for these helpful comments. They have radically changed my overview of Rope!

[Forked issue #3](https://github.com/edreamleo/ekr-fork-rope2/issues/3) will explore these ideas in detail.

**Reply by baggiponte** [reply]
> Type inference is a rather loose description of what rope does, it have some parallels to type inferrence in systems like mypy, but it also isn't really the same thing. Basically, rope needs to know whether a symbol that appears in a code are the same as another symbol, for example for a rename operation. It does some analysis to determine the type of a variable, what attributes/methods they have, etc.

How do pylsp/pyright handle this?

**Reply by lieryan** [reply]
pylsp doesn't do any type inferencing. The libraries that pylsp connects together, like rope (via [pylsp-rope](https://pypi.org/project/pylsp-rope/)) or mypy (via [pylsp-mypy](https://pypi.org/project/pylsp-mypy/)) does all the analysis and type inferencing. But pylsp itself doesn't do any analysis by itself, it just calls libraries like mypy and then translate its internal analysis result format to LSP format, or the refactoring/edit operations done by rope to a JSON containing the diff that describes the change. For the most part, pylsp doesn't care about how the libraries internally does its analysis.

I don't know exactly how mypy works, but for rope, we basically uses the `ast` parser from the standard library to parse python code and then do our analysis on top of it, basically if we see an ast that matches the code that looks like `self.foo = MyClass()`, then we know that `self.foo` have the type of `MyClass` and should have all the methods and attributes from `MyClass`. With type hinted python code, we should also be able to use the type annotations, but currently, type hints is still underutilised by rope.

**Comment by edreamleo** [comment]
@baggiponte I have updated the [readme file](https://github.com/edreamleo/ekr-fork-rope2/tree/devel#readme) to include more information about how I work.

**Comment by lieryan** [comment]
Hi @baggiponte, welcome to Rope. It's always good to have more people interested in supporting and helping with Rope. I'm a Vim user too, so you're in the company of like-minded.

Rope is a community project, so I/we do not set priorities or schedule for any contributors, so you're welcome to work on anything you want at your own pace. Do let me know if you need help with anything, and I will try to help whatever I can.

**Reply by baggiponte** [reply]
Thanks! As far as (neo)vim is concerned, rope is either available as a standalone plugin (ropevim), while if I install python-lsp-server and pylsp-rope I can use it with the native nvim-lspconfig, though it's "very buggy". Is it rope support that is "buggy" or pylsp in general?

**Reply by lieryan** [reply]
The ropevim plugin isn't really actively developed anymore at the moment. Currently, most of the effort at editor integration is on pylsp-rope. 

It's been some time since I've tested Native Neovim LSP, but if you're referring to [my comment in here](https://github.com/python-rope/rope/wiki/Rope-in-Vim-or-Neovim#neovim-native-lsp) and described further in this [ticket](https://github.com/python-rope/pylsp-rope/issues/5), what's buggy is the native Neovim LSP Code Action. 

If you used any other LSP plugin in the Vim ecosystem (vim-lsp, ALE, CoC), everything works just fine with either Vim or Neovim, so I suspect the bug is on Neovim's internal LSP implementation. It's been a while since I tested with the native Neovim LSP, so there's a possibility this has been fixed now.

**Reply by baggiponte** [reply]
> The ropevim plugin isn't really actively developed anymore at the moment. Currently, most of the effort at editor integration is on pylsp-rope.

Makes sense, thanks.

> It's been a while since I tested with the native Neovim LSP, so there's a possibility this has been fixed now.

Let me have a look at that! 💪🏼

**Comment by edreamleo** [comment]
@baggiponte I'm guessing that you may not have to know much (or anything) about type inference.

Rope uses type inference to determine how to do various refactorings. For example, if a refactoring renames an ivar (say `x` in a class `Y`), Rope needs to infer the type of any object whose name is `x` to know whether this object is actually an ivar of `Y`.
